### PR TITLE
Fix Scheduler

### PR DESCRIPTION
**Pull request by adamgfraser**
It appears that the issue with tests in ZIO Test's internal test suite timing out was related to the scheduler. The `TestClock` is powerful enough that it provides its own implementation of the `Scheduler`, but it appears that this was not being used. With these changes all tests are passing. Will highlight a couple of questions / comments below.

### CODE REVIEW

File: core/shared/src/main/scala/zio/scheduler/Scheduler.scala [hunk 1]
```diff
@@ -16,9 +16,17 @@
 
 package zio.scheduler
 
-import zio.{ Has, ZLayer }
+import zio.{ UIO, ZLayer }
 import zio.internal.{ Scheduler => IScheduler }
 
 object Scheduler extends PlatformSpecific {
-  val live: ZLayer.NoDeps[Nothing, Has[IScheduler]] = ZLayer.succeed(globalScheduler)
+  trait Service extends Serializable {
```
Comment (adamgfraser) [inline core/shared/src/main/scala/zio/scheduler/Scheduler.scala]: I think we need `Scheduler` to be a service so that we can express that the `TestClock` implements it.

File: test/shared/src/main/scala/zio/test/environment/TestClock.scala [hunk 2]
```diff
@@ -95,7 +96,7 @@ import zio.clock.Clock
  */
 object TestClock extends Serializable {
 
-  trait Service extends Restorable {
+  trait Service extends Restorable with Clock.Service with Scheduler.Service {
```
Comment (adamgfraser) [inline test/shared/src/main/scala/zio/test/environment/TestClock.scala]: I noticed that the other test services don't extend the live services, which seems to create the possibility that the test service would not implement all the methods of the live service and prevent accessing the live service methods from the test service. Is there a reason for this? I am definitely new to this structure but I would have thought we would want to extend them as above.

File: test/shared/src/main/scala/zio/test/environment/TestClock.scala [hunk 3]
```diff
@@ -386,6 +388,13 @@ object TestClock extends Serializable {
   def setTimeZone(zone: ZoneId): ZIO[TestClock, Nothing, Unit] =
     ZIO.accessM(_.get.setTimeZone(zone))
 
+  /**
+   * Accesses a `TestClock` instance in the environment and returns a new
+   * `Scheduler` backed by this `TestClock`.
+   */
+  val scheduler: ZIO[TestClock, Nothing, IScheduler] =
```
Comment (adamgfraser) [inline test/shared/src/main/scala/zio/test/environment/TestClock.scala]: This is not ideal. I tried to call the `Scheduler.scheduler` method from the package object that I created but then I got an error message that the `Scheduler` was not in the `TestEnvironment` even though I would have thought it would be since the `TestClock` now implements the `Scheduler`. Is there a better way to do this that I am just not understanding?

File: core/shared/src/main/scala/zio/scheduler/Scheduler.scala (hunk 1 above)
Comment (jdegoes) [reply core/shared/src/main/scala/zio/scheduler/Scheduler.scala]: A service can be any value at all. In the old encoding, it was `IScheduler` (`internal.Scheduler`), i.e. an implementation of that interface. 

Why isn't that enough?

File: test/shared/src/main/scala/zio/test/environment/TestClock.scala (hunk 2 above)
Comment (jdegoes) [reply test/shared/src/main/scala/zio/test/environment/TestClock.scala]: In the new scheme, covariance is not supported. That is to say, `Has[Clock.Service] with Has[TestClock.Service]` contains 2 modules, not 1, even if `TestClock.Service <: Clock.Service` (this is necessary for sane reasoning about module updates).

So it makes more sense to factor off the "test methods" from the "clock service", because the `TestClock.live` needs to emit _two_ things: `TestClock.Service` _as well as_ `Clock.Service`.

Does this make sense?

Basically it's an implementation detail that the same class below (`Test`) happens to implement both interfaces, it's not required though.

File: test/shared/src/main/scala/zio/test/environment/TestClock.scala (hunk 3 above)
Comment (jdegoes) [reply test/shared/src/main/scala/zio/test/environment/TestClock.scala]: Actually we can just make `TestClock.live` emit a scheduler too!

Comment (jdegoes) [reply test/shared/src/main/scala/zio/test/environment/TestClock.scala]: To be clear, this would look like:

```scala
live : ZDep[Live, Nothing, Clock with TestClock with Scheduler]
```

The _old_ `Scheduler`, which was a type alias for `Has[IScheduler]`.

**Comment by jdegoes** [comment]
@adamgfraser Superb detective work!

I think the change could be, `Has.allOf[Clock.Service, TestClock.Service, IScheduler](test, test, test)`

Although we have to investigate _why_ the tests were getting any scheduler at all. Where was it coming from, statically? `testEnvironment` will have the clue.

**Comment by adamgfraser** [comment]
I think it is coming from `Clock.live` [here](https://github.com/jdegoes/zio/blob/59af7b377bb95494103ea8b1aae032fdd54be8af/core/shared/src/main/scala/zio/clock/Clock.scala#L35). Because we are building in the live environment and which includes the live clock and that brings a scheduler dependency, if I am understanding correctly. Will work on incorporating your other comments.

**Comment by jdegoes** [comment]
All right, I think I understand now, and here's what I'd do:

Change `ZEnv` to include `Scheduler` (`Has[IScheduler]`). Then the types will force you to compose `TestEnvironment` differently. Then you can change `TestClock.live` to emit another service: `IScheduler`.

**Comment by adamgfraser** [comment]
Working on. I think we still need `Scheduler` to be a trait so that `TestClock.Test` can extend it to fulfill the contract when we do `Has.allOf[Clock.Service, TestClock.Service, Scheduler](test, test, test)`.

**Comment by jdegoes** [comment]
@adamgfraser I'd just make it a field inside `Test`, e.g. `Has.allOf(test, test, test.scheduler)`.

**Comment by adamgfraser** [comment]
Okay, let me try that.

**Comment by jdegoes** [comment]
The old patterns are very verbose and if this new direction is going to work out, it makes sense to eliminate as much ceremony as possible, playing to the strengths of this approach.

**Comment by adamgfraser** [comment]
Ready for another review.

**Comment by jdegoes** [comment]
Looks superb! Separately, I think we can pull out `IScheduler` into `zio.scheduler.Scheduler.Service`, and maybe give it a "safe" method that is public, and make the rest of them private to ZIO package (it occurs to me now we are leaking an impure API).
